### PR TITLE
Draft implementation of file-scoped-namespaces

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Draft impl of championed feature: https://github.com/dotnet/csharplang/issues/137
Speclet at: https://github.com/dotnet/csharplang/pull/4073
Test plan: https://github.com/dotnet/roslyn/issues/49000

To help ground discussion, find issues, and garner feedback to help with the language design process.

The general approach here was to take an approach similar to what we did with TypeScript.  Specifically, given the two namespaces:

`namespace X; class C { }` and
`namespace X { class C { } }`

We represent these with different concrete types, but a similar abstract parse tree shape.  Specifically, both generate a parse tree like so:

```
CompilationUnit
|
NamespaceDeclaration (or SingleLineNamespaceDeclaration)
|
-- Name
|    |
|    -- X
-- Members
    |
    -- ClassDeclaration
        |
       -- Name
            |
            -- C
```

NamespaceDeclaration and the new SingleLineNamespaceDeclaration derive from a common BaseNamespaceDeclaration.  This type contains the Name, Externs, Usings, Attributes and Members.  The two subclasses contain the only syntactic differents (the curlies, or hte semicolon).

By doing this, we allow all code to still presume a certain containment structure for files.  For example, in the IDE there is lots of code that will walk upwards in a file looking for the containing namespace declaration.  This approach allows all existing code (that doesn't care about curlies) to just move to processing 'BaseNamespaceDeclaration's, isntead of 'NamespaceDeclaration's, and allows anything that currently looks for SyntaxKind.NamespaceDeclaration to just look for that and SyntaxKind.SingleLineNamespaceDeclaration.

--

For my testing approach, i made targeted parsing tests and targeted tests for the specific specialized errors we produce.  I then searched the compiler test base for usages of NamespaceDeclarationSyntax, SyntaxKind.NamespaceDeclaration, INamespaceSymbol, and NamespaceSymbol and generally created sibling tests that validated the same behavior when a single-line namespace was used instead.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 1]
```diff
@@ -3051,6 +3060,21 @@
       <Kind Name="SemicolonToken"/>
     </Field>
   </Node>
+  <Node Name="SingleLineNamespaceDeclarationSyntax" Base="BaseNamespaceDeclarationSyntax">
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3063]: By convention we call these "directive" I suggest `NamespaceDirectiveSyntax`

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -368,12 +403,12 @@ private enum NamespaceParts
             TopLevelStatementsAfterTypesAndNamespaces = 6,
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:403]: don't we need to modify `NamespaceParts`?

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:403]: in what way?

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3063]: i don't actually like that.  because now you'd have:

```
BaseNamespaceDeclarationSyntax
|--NamespaceDeclarationSyntax
|--NamespaceDirectiveSyntax
```

We pick 'directive' for things that are very different (i.e. using statement vs directive).  In this case, these are the practically the same.  It's really a specialization of a core concept.  So i think the name should be SomethingNamespaceDeclaration.  Now, if that's SingleLineNamespaceDeclaration, or FileScopedNamespaceDeclaration, or something else is open for me.  But the latter part i believe is named properly :)

Comment (alrz) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3063]: Maybe call the base BaseNamespaceSyntax.

This is a directive as it affects whatever comes after, it is not a declaration in and of itself, there are a few other nodes in the language already with this semantic to make a comparison.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:403]: If we allow the same ordering of regular namepsaces as you clarified in the other thread we probably don't need to.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3063]: i can def discuss this with the team.  Having:

```
NamepsaceSyntax
|--NamespaceDeclarationSyntax
|--NamespaceDirectiveSyntax
```

Is potentially possible.  I still dont' personally like it though.  I still see these as declarations, even if there is no body.  In other words, a declaration is not defined by having a body, but by defining a symbol (or symbols).

Comment (alrz) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3063]: >but by defining a symbol

I agree, I think you can look at it either way, it is a declaration as it defines a symbol but a directive syntactically. I think that hierarchy is expressive enough though and we don't have to think of another term to explain it.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 3]
```diff
@@ -3027,20 +3027,29 @@
       </PropertyComment>
     </Field>
   </AbstractNode>
-  <Node Name="NamespaceDeclarationSyntax" Base="MemberDeclarationSyntax">
+    <AbstractNode Name="BaseNamespaceDeclarationSyntax" Base="MemberDeclarationSyntax">
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3030]: ><AbstractNode [](start = 4, length = 13)

There's an indentation difference with rest of XML file

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 4]
```diff
@@ -758,7 +758,7 @@ public override Binder VisitInterfaceDeclaration(InterfaceDeclarationSyntax node
             public override Binder VisitRecordDeclaration(RecordDeclarationSyntax node)
                 => VisitTypeDeclarationCore(node);
 
-            public override Binder VisitNamespaceDeclaration(NamespaceDeclarationSyntax parent)
+            public sealed override Binder VisitNamespaceDeclaration(NamespaceDeclarationSyntax parent)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:761]: >sealed [](start = 19, length = 6)

no need to add `sealed`, as the type is sealed already

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 5]
```diff
@@ -774,7 +774,23 @@ public override Binder VisitNamespaceDeclaration(NamespaceDeclarationSyntax pare
...
 
-            internal Binder VisitNamespaceDeclaration(NamespaceDeclarationSyntax parent, int position, bool inBody, bool inUsing)
+            public override Binder VisitSingleLineNamespaceDeclaration(SingleLineNamespaceDeclarationSyntax parent)
+            {
+                if (!LookupPosition.IsInNamespaceDeclaration(_position, parent))
+                {
+                    return VisitCore(parent.Parent);
+                }
+
+                // test for position equality in case the open brace token is missing:
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:784]: comment needs to be adjusted

File: src/Compilers/CSharp/Portable/Binder/Imports.cs [hunk 6]
```diff
@@ -83,9 +83,10 @@ internal string GetDebuggerDisplay()
                 usingDirectives = inUsing ? default(SyntaxList<UsingDirectiveSyntax>) : compilationUnit.Usings;
                 externAliasDirectives = compilationUnit.Externs;
             }
-            else if (declarationSyntax.Kind() == SyntaxKind.NamespaceDeclaration)
+            else if (declarationSyntax.Kind() == SyntaxKind.NamespaceDeclaration ||
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Imports.cs:86]: >declarationSyntax.Kind() == SyntaxKind.NamespaceDeclaration || [](start = 21, length = 62)

nit: `declarationSyntax.Kind() is (SyntaxKind.NamespaceDeclaration or SyntaxKind.SingleLineNamespaceDeclaration)`
also applies in other places in this PR

**Comment by jcouv** [comment]
Done with review pass (iteration 19), not looking at the tests much yet. Opened test plan https://github.com/dotnet/roslyn/issues/49000

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:761]: ah.  this was still helpful to me to know that there wasn't some subclass i'd have to worry about.  i can remove this though.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:784]: yup!

File: src/Compilers/CSharp/Portable/Binder/Imports.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/Imports.cs:86]: good call :)

**Comment by jaredpar** [comment]
From the title I thought this was more experimentation but given that reviews are happening it's starting to feel like the intent is more to implement and ship the feature here. If that is the case then we should really begin following our normal process here: create a feature branch, assign a dev, schedule a secondary reviewer, etc ... and breaking this up into digestible PRs. 

What I really want to avoid is this turning into a very large PR that implements a language feature in the entirety. Those are generally viewed negatively as dev's don't want to devote multiple days of time to review a single PR (or even a single day). This PR isn't there yet but also unsure how big this will end up being (gut feel is this is bigger than a single PR).

**Comment by CyrusNajmabadi** [comment]
> From the title I thought this was more experimentation 

This was more for experimentation.  I wanted to be able to come in with an idea of the costs here when proposing/designing the lang feature and discussing hte impl :)

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml]: fixed.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:784]: fixed.

**Comment by CyrusNajmabadi** [comment]
Taking out of draft.  This PR implements https://github.com/dotnet/csharplang/pull/4073 and follows the latest decisions from the LDM.

@jaredpar @jcouv @jinujoseph how should i proceed here?  I would def love a compiler buddy who could help me get this in.  thanks!

**Comment by jaredpar** [comment]
@chsienki was going to take this over compiler side. Our plan was to create a branch, and start moving froward from there.

**Comment by CyrusNajmabadi** [comment]
> @chsienki was going to take this over compiler side. Our plan was to create a branch, and start moving froward from there.

Sgtm.  @chsienki do you want to schedule some time with me to go over how this works here?

**Comment by chsienki** [comment]
@CyrusNajmabadi lets sync up offline and work out how to get this in.

**Comment by chsienki** [comment]
@CyrusNajmabadi Lol, beat me to it. Will do.

**Comment by jcouv** [comment]
Changed PR to draft for now. Please ping when we should take another look.

**Comment by CyrusNajmabadi** [comment]
> Changed PR to draft for now. Please ping when we should take another look.

?

This is ready for review.  That's why i brought it out of draft and pinged all of you :)

**Comment by jaredpar** [comment]
> This is ready for review. That's why i brought it out of draft and pinged all of you

Before we can review this we need to decide on how, where we're ingesting in. Features are pretty much never taken directly into master but instead go through a [feature branch](https://github.com/dotnet/roslyn/blob/master/docs/contributing/Developing%20a%20Language%20Feature.md#process). It's generally just not possible to get all of the feature work into a reasonbly sized PR. Only features which border on bug fixes can be done this way. 

@chsienki is shepparding this compiler side. He's going to figure out the branching, get the PR targeted there and then we can move on with reviewing the code .

**Comment by chsienki** [comment]
Re-based onto the feature branch.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 7]
```diff
@@ -2542,7 +2582,7 @@ private bool IsMisplacedModifier(SyntaxListBuilder modifiers, SyntaxList<Attribu
                 {
                     result = incompleteMember;
                 }
-                else if (parentKind == SyntaxKind.NamespaceDeclaration ||
+                else if (parentKind is SyntaxKind.NamespaceDeclaration or SyntaxKind.SingleLineNamespaceDeclaration ||
                          parentKind == SyntaxKind.CompilationUnit && !IsScript)
                 {
                     result = this.AddErrorToLastToken(incompleteMember, ErrorCode.ERR_NamespaceUnexpected);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2588]: Please add a test for this case

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2588]: i think i do.  but will ensure i do.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 8]
```diff
@@ -4470,6 +4510,7 @@ private void ParseVariableDeclarators(TypeSyntax type, VariableFlags flags, Sepa
             // the reported errors should take into consideration whether or not one expects them in the current context.
             bool variableDeclarationsExpected =
                 parentKind != SyntaxKind.NamespaceDeclaration &&
+                parentKind != SyntaxKind.SingleLineNamespaceDeclaration &&
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4513]: Is there a test for this change?

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 9]
```diff
@@ -682,7 +682,7 @@
     <value>'{0}': no suitable method found to override</value>
   </data>
   <data name="ERR_NamespaceUnexpected" xml:space="preserve">
-    <value>A namespace cannot directly contain members such as fields or methods</value>
+    <value>A namespace cannot directly contain members such as fields, methods or statements</value>
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:685]: This reads as statements being a kind of member

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:685]: what's a better term?  :)

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:685]: technically, this is how we represent it, GlobalStatementSyntax wraps a statement into a MemberDecl so we can have it in the compilation-unit/namespace-decl.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 10]
```diff
@@ -274,19 +280,48 @@ internal CompilationUnitSyntax ParseCompilationUnitCore()
...
-                this.ParseNamespaceBody(ref openBrace, ref body, ref initialBadNodes, SyntaxKind.NamespaceDeclaration);
+                if (openBrace == null)
+                {
+                    Debug.Assert(semicolon != null);
 
-                var closeBrace = this.EatToken(SyntaxKind.CloseBraceToken);
-                var semicolon = this.TryEatToken(SyntaxKind.SemicolonToken);
+                    SyntaxListBuilder initialBadNodes = null;
+                    this.ParseNamespaceBody(ref semicolon, ref body, ref initialBadNodes, SyntaxKind.SingleLineNamespaceDeclaration);
+                    Debug.Assert(initialBadNodes == null); // init bad nodes should have been attached to open brace...
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:294]: Comment is wrong (would be attached to semicolon). Do we have a test scenario where that happens?

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs [hunk 11]
```diff
@@ -401,6 +401,25 @@ public void NamespaceWithoutName()
...
+        [WorkItem(539740, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/539740")]
+        [Fact]
+        public void SingleLineNamespaceWithoutName()
+        {
+            var text = "namespace;";
+            var tree = Parse(text);
+            var comp = CreateCompilation(tree);
+            var model = comp.GetSemanticModel(tree);
+            var errors = comp.GetDiagnostics().ToArray();
+            Assert.Equal(2, errors.Length);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs:413]: nit: Let's check the diagnostics instead of just the count

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs [hunk 12]
```diff
@@ -7056,6 +7078,26 @@ class c
...
+";
+            var root = SyntaxFactory.ParseSyntaxTree(source).GetRoot();
+
+            Assert.Equal(source, root.ToFullString());
+            // Verify incomplete class decls don't eat tokens of surrounding nodes
+            var classDecl = root.DescendantNodes().OfType<ClassDeclarationSyntax>().Single();
+            Assert.False(classDecl.Identifier.IsMissing);
+            Assert.True(classDecl.OpenBraceToken.IsMissing);
+            Assert.True(classDecl.CloseBraceToken.IsMissing);
+            var ns = root.DescendantNodes().OfType<SingleLineNamespaceDeclarationSyntax>().Single();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:7098]: Did you intend to check something on `ns`?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 26) with a couple test suggestions

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs [hunk 13]
```diff
@@ -298,6 +306,57 @@ public override SingleNamespaceOrTypeDeclaration VisitNamespaceDeclaration(Names
...
+            {
+                Debug.Assert(node is NamespaceDeclarationSyntax);
+
+                //      namespace X.Y;
+                //      namespace A.B { }
+                if (node.Parent is SingleLineNamespaceDeclarationSyntax)
+                {
+                    diagnostics.Add(ErrorCode.ERR_SingleLineAndNormalNamespace, node.Name.GetLocation());
+                }
+            }
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:358]: We should also have a check to disallow top-level statements.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/DeclaringSyntaxNodeTests.cs [hunk 14]
```diff
@@ -421,28 +464,27 @@ namespace System {}
         }
 
         [Fact]
-        public void TypeParameterDeclaringSyntax()
+        public void TypeParameterDeclaringSyntaxSingleLineNamespace()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/DeclaringSyntaxNodeTests.cs:467]: I didn't understand why the "regular" namespace version of the test was essentially removed here.

File: src/Compilers/CSharp/Test/Symbol/Symbols/AssemblyAndNamespaceTests.cs [hunk 15]
```diff
@@ -471,5 +625,166 @@ namespace N { }
                 // (2,1): error CS1671: A namespace declaration cannot have modifiers or attributes
                 Diagnostic(ErrorCode.ERR_BadModifiersOnNamespace, "[System.Obsolete]").WithLocation(1, 1));
         }
+
+        private static readonly CSharpParseOptions s_previewOptions = CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Preview);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Symbol/Symbols/AssemblyAndNamespaceTests.cs:629]: nit: Typically we use `TestOptions.RegularPreview`

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs [hunk 16]
```diff
@@ -28,13 +28,15 @@ internal sealed class FileChangeWatcher
...
-        /// the end of the queue.
+        /// We create a queue of tasks against the IVsFileChangeEx service for two reasons. First, we are obtaining the
+        /// service asynchronously, and don't want to block on it being available, so anybody who wants to do anything
+        /// must wait for it. Secondly, the service itself is single-threaded; the entry points are asynchronous so we
+        /// avoid starving the thread pool, but there's still no reason to create a lot more work blocked than needed.
+        /// Finally, since this is all happening async, we generally need to ensure that an operation that happens for
+        /// an earlier call to this is done before we do later calls. For example, if we started a subscription for a
+        /// file, we need to make sure that's done before we try to unsubscribe from it.  For performance and
+        /// correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block
+        /// and add any work to the end of the queue.
```
Comment (RikkiGibson) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:39]: I didn't understand why this change was included in the PR.

**Review by RikkiGibson — APPROVED** [review]
LGTM. My expectation is I will make the changes to resolve the comments myself in a follow-up PR.

**Comment by RikkiGibson** [comment]
@jcouv @chsienki My expectation is that we will try to merge this PR more or less as-is, then submit a follow up which resolves any unresolved comments on this PR. Does that seem reasonable? Or should I maybe just push commits to this PR?

**Comment by jaredpar** [comment]
@RikkiGibson That seems reasonable to me. This PR is already quite large, a bit unweildy (evidenced by the current conflicts), and we are going into a feature branch. I would focus on getting this merged then follow up with a PR that cleans up the items that you noticed.

File: src/Compilers/CSharp/Portable/Binder/Imports.cs [hunk 17]
```diff
@@ -83,9 +83,9 @@ internal string GetDebuggerDisplay()
                 usingDirectives = inUsing ? default(SyntaxList<UsingDirectiveSyntax>) : compilationUnit.Usings;
                 externAliasDirectives = compilationUnit.Externs;
             }
-            else if (declarationSyntax.Kind() == SyntaxKind.NamespaceDeclaration)
+            else if (declarationSyntax.Kind() is SyntaxKind.NamespaceDeclaration or SyntaxKind.SingleLineNamespaceDeclaration)
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Imports.cs:86]: @RikkiGibson I find 5 or 6 instances of this check so an extension sounds fine. I'd say push a commit if you'd like (you're in charge of the PR ;-)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/DeclaringSyntaxNodeTests.cs (hunk 14 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/DeclaringSyntaxNodeTests.cs:467]: Good point. Let's add the original back, or make this a Theory to cover both

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs (hunk 16 above)
Comment (jcouv) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:39]: I assumed this is an artifact of having a merge PE (from `main` branch) included in the PR.
But it looks like I was wrong. This change came from a commit called "Get building" ([https://github.com/dotnet/roslyn/commit/a183d9e50a4fc18f1796e59ce8535eeb28c06e04](https://github.com/dotnet/roslyn/commit/a183d9e50a4fc18f1796e59ce8535eeb28c06e04)) which is new to this PR. 
Let's revert it.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs:39]: i can't explain it either.  feel free to revert.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 28)

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs (hunk 13 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:382]: @RikkiGibson  Add PROTOTYPE or track elsewehere?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4598]: @RikkiGibson to ensure we've addressed this

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:293]: @RikkiGibson to address

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs (hunk 12 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:7098]: @RikkiGibson to address

File: src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs (hunk 13 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Declarations/DeclarationTreeBuilder.cs:382]: I think this is handled by the existing mechanisms that prevent top-level statements within namespaces. See `SingleLineNamespaceWithFollowingStatement` in `AssemblyAndNamespaceTests.cs`

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs:413]: I am OK with this because it's taken from the test directly above.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs (hunk 12 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:7098]: It's not that meaningful but I'll add an assert that the semicolon is present.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:293]: See `SingleLineDeclarationParsingTests.NamespaceWithExtraSemicolon`. Will fix the comment.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4598]: Adding the tests.

File: src/Compilers/CSharp/Portable/Binder/Imports.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Imports.cs]: Not really interested in doing this at this time. Would rather make the adjustment as part of a larger refactoring e.g. to replace `Kind()` tests with type tests in more places.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 18]
```diff
@@ -2266,7 +2268,8 @@ private NamespaceOrTypeSymbol GetDeclaredTypeMemberContainer(CSharpSyntaxNode me
             if (memberDeclaration.Parent.Kind() == SyntaxKind.CompilationUnit)
             {
                 // top-level namespace:
-                if (memberDeclaration.Kind() == SyntaxKind.NamespaceDeclaration)
+                if (memberDeclaration.Kind() == SyntaxKind.NamespaceDeclaration ||
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:2281]: Done

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 19]
```diff
@@ -2297,7 +2300,8 @@ private NamespaceOrTypeSymbol GetDeclaredTypeMemberContainer(CSharpSyntaxNode me
             }
 
             // a namespace or a type in an explicitly declared namespace:
-            if (memberDeclaration.Kind() == SyntaxKind.NamespaceDeclaration || SyntaxFacts.IsTypeDeclaration(memberDeclaration.Kind()))
+            if (memberDeclaration.Kind() is SyntaxKind.NamespaceDeclaration or SyntaxKind.SingleLineNamespaceDeclaration ||
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:2313]: Done

File: src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs [hunk 20]
```diff
@@ -4468,6 +4495,99 @@ class C
                 SymbolDisplayPartKind.EventName);
         }
 
+        [WorkItem(791756, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/791756")]
+        [Fact]
+        public void KindOptionsSingleLineNamespace()
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:4500]: Have done this for many of the tests.

File: src/Compilers/CSharp/Test/Symbol/Symbols/AssemblyAndNamespaceTests.cs (hunk 15 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/AssemblyAndNamespaceTests.cs:629]: Renamed this. I think it could be handy to be able to change this from Preview to Regular in one place. We have done this for other features in the past, IIRC.
